test/object-store/object.cpp

```diff
@@ -2161,3 +2165,75 @@ TEST_CASE("Embedded Object") {
         REQUIRE(calls == 2);
     }
 }
+
+#if REALM_ENABLE_AUTH_TESTS
+
+TEST_CASE("Asymmetric Object") {
+    Schema schema{
+        {"asymmetric",
+         ObjectSchema::IsAsymmetric{true},
+         {
+             {"_id", PropertyType::Int, Property::IsPrimary{true}},
+             {"location", PropertyType::Int},
+             {"reading", PropertyType::Int},
+         }},
+        {"asymmetric_link",
+         ObjectSchema::IsAsymmetric{true},
+         {
+             {"_id", PropertyType::Int, Property::IsPrimary{true}},
+             {"location", PropertyType::Mixed | PropertyType::Nullable},
+         }},
+        {"table",
+         {
+             {"_id", PropertyType::Int, Property::IsPrimary{true}},
+             {"location", PropertyType::Int},
+             {"reading", PropertyType::Int},
+         }},
+    };
+
+    realm::app::FLXSyncTestHarness harness("asymmetric_sync", {schema});
+    SyncTestFile config(harness.app()->current_user(), schema, SyncConfig::FLXSyncEnabled{});
+
+    auto realm = Realm::get_shared_realm(config);
+    CppContext ctx(realm);
+
+    auto create = [&](util::Any&& value, std::string table_name, CreatePolicy policy = CreatePolicy::ForceCreate) {
+        realm->begin_transaction();
+        auto obj = Object::create(ctx, realm, *realm->schema().find(table_name), value, policy);
+        realm->commit_transaction();
+        return obj;
+    };
+
+    SECTION("Basic object creation") {
+        auto obj = create(
+            AnyDict{
+                {"_id", INT64_C(1)},
+                {"location", INT64_C(10)},
+                {"reading", INT64_C(20)},
+            },
+            "asymmetric");
+        // Object returned is not valid.
+        REQUIRE(!obj.obj().is_valid());
+        // Object gets deleted immediately.
+        REQUIRE(Results(realm, realm->read_group().get_table("class_table")).size() == 0);
+    }
```

Comment:
Ideally we would have some basic integration tests:
- make sure that the server does persist the object created by the client
- a second client does not sync down any objects created on an asymmetric table
- how does the server react when two clients create an asymmetric object with the same pk?
- is there a sensible sync error message when a client opens a table as normal top level type, but it is asymmetric type on the server?

Reply:
good suggestions. I have some of these tests but I didn't push them yet.